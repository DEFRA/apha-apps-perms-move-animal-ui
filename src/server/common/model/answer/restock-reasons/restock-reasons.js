import { CheckboxAnswer } from '../checkbox/checkbox.js'

/** @import {CheckboxConfig, CheckboxData} from '../checkbox/checkbox.js' */

/**
 * @typedef {{ restockReasons: CheckboxData }} RestockReasonPayload
 */

/**
 * @augments {CheckboxAnswer<RestockReasonPayload>}
 */
export class RestockReasonsAnswer extends CheckboxAnswer {
  /** @type {CheckboxConfig} */
  static config = {
    payloadKey: 'restockReasons',
    isPageHeading: true,
    hint: 'Select all that apply.',
    options: {
      fattening: {
        label: 'Fattening'
      },
      breeding: {
        label: 'Breeding'
      },
      sucklingRestocking: {
        label: 'Suckling restocking'
      },
      dairyRestocking: {
        label: 'Dairy restocking'
      },
      other: {
        label: 'Something else'
      }
    },
    errors: {
      emptyOptionText: {
        message: 'Select the reasons you have for restocking'
      }
    }
  }
}
