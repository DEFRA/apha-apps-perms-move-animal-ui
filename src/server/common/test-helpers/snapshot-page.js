import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
