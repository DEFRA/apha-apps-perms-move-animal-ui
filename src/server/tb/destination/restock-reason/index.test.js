import { RestockReasonsAnswer } from '../../../common/model/answer/restock-reasons/restock-reasons.js'
import { describePageSnapshot } from '../../../common/test-helpers/snapshot-page.js'
import { RestockReasonPage, restockReasonPage } from './index.js'
import { additionalInfoPage } from '../additional-info/index.js'
import { restockAdditionalInfoPage } from '../restock-additional-info/index.js'

const sectionKey = 'destination'
const question = 'Which reasons do you have for restocking?'
const questionKey = 'restockReasons'
const pageUrl = '/destination/restocking-additional-info-reason'

describe('RestockReasonPage', () => {
  const page = new RestockReasonPage()

  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct question', () => {
    expect(page.question).toBe(question)
  })

  it('should have the correct questionKey', () => {
    expect(page.questionKey).toBe(questionKey)
  })

  it('should have the correct Answer model', () => {
    expect(page.Answer).toBe(RestockReasonsAnswer)
  })

  it('nextPage should return biosecurity summary page when the answer does not include "other"', () => {
    const answer = new RestockReasonsAnswer({
      restockReasons: ['fattening']
    })
    const nextPage = page.nextPage(answer)
    expect(nextPage).toBe(additionalInfoPage)
  })

  it('nextPage should return other wildlife measures page when the answer includes "other"', () => {
    const answer = new RestockReasonsAnswer({
      restockReasons: ['fattening', 'other']
    })
    const nextPage = page.nextPage(answer)
    expect(nextPage).toBe(restockAdditionalInfoPage)
  })

  it('should export page', () => {
    expect(restockReasonPage).toBeInstanceOf(RestockReasonPage)
  })

  describePageSnapshot({
    describes: 'restockReasonPage.content',
    it: 'should render expected response and content',
    pageUrl
  })
})
