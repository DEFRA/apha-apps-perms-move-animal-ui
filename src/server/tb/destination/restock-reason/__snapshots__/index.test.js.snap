// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestockReasonPage restockReasonPage.content should render expected response and content 1`] = `
"
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<div class="govuk-form-group">
  <fieldset class="govuk-fieldset" aria-describedby="restockReasons-hint">
  <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
    <h1 class="govuk-fieldset__heading">
      Which reasons do you have for restocking?
    </h1>
  </legend>
  <div id="restockReasons-hint" class="govuk-hint">
    Select all that apply.
  </div>
  <div class="govuk-checkboxes" data-module="govuk-checkboxes">
    <div class="govuk-checkboxes__item">
      <input class="govuk-checkboxes__input" id="restockReasons" name="restockReasons" type="checkbox" value="fattening" data-testid="fattening-checkbox">
      <label class="govuk-label govuk-checkboxes__label" for="restockReasons">
        Fattening
      </label>
    </div>
    <div class="govuk-checkboxes__item">
      <input class="govuk-checkboxes__input" id="restockReasons-2" name="restockReasons" type="checkbox" value="breeding" data-testid="breeding-checkbox">
      <label class="govuk-label govuk-checkboxes__label" for="restockReasons-2">
        Breeding
      </label>
    </div>
    <div class="govuk-checkboxes__item">
      <input class="govuk-checkboxes__input" id="restockReasons-3" name="restockReasons" type="checkbox" value="sucklingRestocking" data-testid="sucklingRestocking-checkbox">
      <label class="govuk-label govuk-checkboxes__label" for="restockReasons-3">
        Suckling restocking
      </label>
    </div>
    <div class="govuk-checkboxes__item">
      <input class="govuk-checkboxes__input" id="restockReasons-4" name="restockReasons" type="checkbox" value="dairyRestocking" data-testid="dairyRestocking-checkbox">
      <label class="govuk-label govuk-checkboxes__label" for="restockReasons-4">
        Dairy restocking
      </label>
    </div>
    <div class="govuk-checkboxes__item">
      <input class="govuk-checkboxes__input" id="restockReasons-5" name="restockReasons" type="checkbox" value="other" data-testid="other-checkbox">
      <label class="govuk-label govuk-checkboxes__label" for="restockReasons-5">
        Something else
      </label>
    </div>
  </div>
</fieldset>
</div>

            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
